### PR TITLE
Add a visibility icon to status

### PR DESCRIPTION
**Pull request by noellabo**
Here's another implementation that deals with the same problem as https://github.com/tootsuite/mastodon/pull/14119

- Add a visibility icon to the top right (top left in RTL) of status
- It doesn't add anything to the public status (change as little as possible)
- It's looking for a balance between being fully identifiable and unobtrusive
- It uses CSS to describe the look and feel, so it's easy to change it in the theme
- Added some missing classes (you can even change the color of the entire status, depending on the theme)

On the timeline
visibility-icon_a

View Details
visibility-icon_b

### CODE REVIEW

File: app/javascript/styles/mastodon/components.scss [hunk 1]
```diff
@@ -1052,6 +1052,29 @@
...
+      content: "\f09C";
+  }
+
+  &-private .status__relative-time::before {
+      content: "\F023";
+  }
+
+  &-direct .status__relative-time::before {
+      content: "\F0E0";
+  }
```
Comment (ThibG) [inline app/javascript/styles/mastodon/components.scss:1077]: I'm a bit uneasy relying on “internal” FontAwesome values like this.

**Review by ClearlyClaire — APPROVED** [review]
That's a change I've been wanting for a long time, thanks!
I am not completely sure I like the 100%-CSS approach, especially with directly using FontAwesome's “internal” values, but… that thing aside, it looks good to me.

**Comment by Gargron** [comment]
Two comments:

- It would be better to use an icon element with a `title` attribute for a tooltip instead of only CSS
- I am not sure of the value of having the icon next to the name in detailed view

Would it not be better to have the icon next to the timestamp there also?

File: app/javascript/styles/mastodon/components.scss (hunk 1 above)
Comment (Sasha-Sorokin) [reply app/javascript/styles/mastodon/components.scss:1077]: I agree! Please do not do this, this is not accessible by any means. The content of pseudo elements is being read by screenreaders and can lead to the confusion, instead you might want to place span with icon (`<i>`), inside or before the relative time in the code of status component, and provide `aria-label` to it describing privacy level for people, who unable to see the icon.

**Comment by Sasha-Sorokin** [comment]
Don't we already have privacy level in detailed view already? Might be not for unlisted posts, but private and direct ones for sure. Probably just should make it to show public and unlisted ones too. And also, while we onto this, it would be nice to also remove replacement of the boost button when it is not available, just rather mark it inactive, not replace with :lock: :email: (it wouldn't make sense to duplicate privacy level and it is not very accessible right now, see #14088). I previously expressed my feelings that I really like this conception, thanks for taking time to implement it.

**Comment by noellabo** [comment]
- Changed to use the icon element with a title attribute.
- Detaild view now shows an icon next to the timestamp.
- Detaild view has been rearranged to not display two icons, private and direct.

visibility-icon_c
visibility-icon_d

File: app/helpers/application_helper.rb [hunk 2]
```diff
@@ -77,6 +77,16 @@ def fa_icon(icon, attributes = {})
...
   end
 
+  def visibility_icon(status)
+    if status.unlisted_visibility?
+      fa_icon('unlock', title: I18n.t('statuses.visibilities.unlisted'))
+    elsif status.private_visibility? || status.limited_visibility?
+      fa_icon('lock', title: I18n.t('statuses.visibilities.private'))
+    elsif status.direct_visibility?
+      fa_icon('envelope', title: I18n.t('statuses.visibilities.direct'))
+    end
```
Comment (ThibG) [inline app/helpers/application_helper.rb:87]: I would also display the globe for public toots.

File: app/javascript/mastodon/components/status.js [hunk 3]
```diff
@@ -417,6 +423,15 @@ class Status extends ImmutablePureComponent {
       statusAvatar = <AvatarOverlay account={status.get('account')} friend={account} />;
     }
 
+    const visibilityIconInfo = {
+      'public': null,
```
Comment (ThibG) [inline app/javascript/mastodon/components/status.js:427]: Same here, I would also display the globe.

File: app/javascript/mastodon/features/status/components/detailed_status.js [hunk 4]
```diff
@@ -158,34 +166,47 @@ export default class DetailedStatus extends ImmutablePureComponent {
...
-      applicationLink = <span> · <a className='detailed-status__application' href={status.getIn(['application', 'website'])} target='_blank' rel='noopener noreferrer'>{status.getIn(['application', 'name'])}</a></span>;
+      applicationLink = <React.Fragment> · <a className='detailed-status__application' href={status.getIn(['application', 'website'])} target='_blank' rel='noopener noreferrer'>{status.getIn(['application', 'name'])}</a></React.Fragment>;
     }
 
-    if (status.get('visibility') === 'direct') {
-      reblogIcon = 'envelope';
-    } else if (status.get('visibility') === 'private') {
-      reblogIcon = 'lock';
+    const visibilityIconInfo = {
+      'public': null,
```
Comment (ThibG) [inline app/javascript/mastodon/features/status/components/detailed_status.js:173]: Same.

**Review by ClearlyClaire — APPROVED** [review]
Looks good to me, but I would also display an icon for public toots.

**Review by ClearlyClaire — APPROVED** [review]

**Comment by Sasha-Sorokin** [comment]
Should we display it though? I really don't see globe on every post being useful. “Unlisted”, “private” and “direct” are being unusual privacy levels and may have unwanted effects, so it has sense to display them, but the globe on every post? I dunno.

**Comment by ThibG** [comment]
@Sasha-Sorokin I don't think that public being the “default” makes much sense. I post in unlisted by default, and my TL is a pretty even mix of all visibilities. Also we've been displaying all of them in glitch-soc and i don't think this has ever been an issue.

**Comment by Sasha-Sorokin** [comment]
Also, do you think it's a good idea to place icon before the relative time rather than after it? In Russian locale we have a little different short forms of “m” (“мин”) and “h” (“ч”), therefore this will cause the icon to move. If we place it after the relative time it should stay at the same position at all times.

**Comment by noellabo** [comment]
I like omitting the globe icon in public status.

Because unlisted and private are conspicuous and easy to distinguish.

Also, in the LTL and FTL, the globe icon is displayed on all posts, making it feel noisy.

However, the globe icon has a title attribute, which clearly explains that the status is public. This should be useful for screen readers etc.

The globe can be changed just by reverting the last commit, so please choose one.

**Comment by ThibG** [comment]
You've got a good point for the public TL but I've never seen such a complaint on glitch-soc so I don't think it looks so noisy…

@Sasha-Sorokin hm… good point as well, it probably would be better to make sure they keep being aligned

**Comment by Sasha-Sorokin** [comment]
Probably bothering too much, sorry. I just noticed that #14132 is going to address #14088 (it says `Fix #...`).

Does this PR fix (some or all of) these:  #1564, #5055, #9788, #13056? If so, may you include them in PR body here too, so users subscribed to previous issues would be notified about this change? 🙂

**Comment by mayaeh** [comment]
I don't think it's a good idea to just use a shade of color to indicate whether a post can be "boost".

I also don't think we need to show the globe icon on every public post.

**Comment by noellabo** [comment]
To improve the appearance of disabled, let's make a new issue or pull request.
If you don't care about fixing the position, you can hide the buttons that cannot be used.

**Comment by Sasha-Sorokin** [comment]
> @mayaeh 
> I don't think it's a good idea to just use a shade of color to indicate whether a post can be "boost".

This post will add indication of the privacy level and we really don't need to duplicate this indication on the boost button. It also seems to be a bad design to have other icons for the single button, #14132 will provide title to describe why the boost action is unavailable. I believe many users are aware that semi-transparent button means it is disabled, and thanks to the `title` they will be reassured of that.

**Comment by noellabo** [comment]
The globe on the timeline can be hidden with CSS if you like. That should be enough.
```
.status__visibility-icon .fa-globe {display: none;}
```

If the current "disabled" is not enough to indicate that you cannot boost, you may want to use a more inconspicuous color or remove it. (https://github.com/tootsuite/mastodon/pull/14132)
```
.icon-button.disabled {opacity: 0.4;}
```
```
.icon-button.disabled {display: none;}
```
visibility-icon_i

A discussion of how it is best to represent a button when it is disabled should be dealt with in a different issue.

Is there anything else you have to do?

**Comment by blank71** [comment]
Two opinion.

I'm disagree to make shaded boost botton to mean that can't boost. Actually, the shaded boost botton is not good for distinguishing that can boost or not. In the light theme, the visiability can be say not bad, but in the dark theme, it's worse. I think, add diagonal line to boost botton to mean that can't boost is better way.

I can understand that add global icon to Pulic toot to make attribute more explicitly. Other side it would be noisy at LTL or FTL, I stand by this opinion. Just disable global icon at LTL and FTL, enable it at HTL, is one way I thought, but it isn't good way, is it?

**Comment by MightyPork** [comment]
incredible that this took six years to be ported from glitch-soc, but still, nice :clap:

**Pull request merged**